File: Sorting/mergeSort.js
/**
Time Complexity
===============
Best Case: O(nlog n)

Average Case: O(nlog n)

Worst Case: O(nlog n)

Space Complexity => O(n)

Divides the array into single element arrays, merges them by sorting one step at a time.
 */

function mergeSort(arr) {
    if (arr.length <= 1) return arr;
    let middle = Math.floor(arr.length / 2);
    let left = mergeSort(arr.slice(0, middle));
    let right = mergeSort(arr.slice(middle));
    return merge(left, right);
}

function merge(left, right) {
    let mergedArr = [];
    let i = 0;
    let j = 0;
    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            mergedArr.push(left[i]);
            i++;
        } else {
            mergedArr.push(right[j]);
            j++;
        }
    }
    while (i < left.length) {
        mergedArr.push(left[i]);
        i++;
    }
    while (j < right.length) {
        mergedArr.push(right[j]);
        j++;
    }
    return mergedArr;
}

console.log(mergeSort([2, 4, 6, 3, 2, 5, 76, 88]));

File: Fundamentals/Sorting/mergeSort.py
def mergeSort(arr):
    if len(arr) <= 1:
        return arr
    middle = int(round(len(arr) / 2))
    left = mergeSort(arr[0: middle])
    right = mergeSort(arr[middle:])
    return merge(left, right)


def merge(left, right):
    mergedArr = []
    i = 0
    j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            mergedArr.append(left[i])
            i += 1
        else:
            mergedArr.append(right[j])
            j += 1
    while i < len(left):
        mergedArr.append(left[i])
        i += 1
    while j < len(right):
        mergedArr.append(right[j])
        j += 1
    return mergedArr


print(mergeSort([100, 56, 49, 50, 56, 2, 3, 5, 1, -16, 0, 6, 7, 101]))
